### ISSUE TITLE
libxi (xi-core as a library)

### ISSUE THREAD

**Issue by cmyr** [google/xi-editor#439]
Breaking this out of an irc conversation from a couple days ago, which started [here](https://botbot.me/mozilla/xi/2017-11-21/?msg=93777902&page=1). The basic gist is that we are going to move away from the current multi-process architecture and towards the more traditional approach of treating xi-core as a library that clients will link against at compile time.

We would still be communicating over JSON, but using the C ABI instead of stdio.

[raph:](https://botbot.me/mozilla/xi/2017-11-21/?msg=93787342&page=2)
>i think keep the current interface, basically
> the C interface would basically be send_to_core(char* json_data, size_t length)
> and then I _think_ preceded by start_core(void (*from_core_callback)(char* json_data, size_t length))

**Comment by eyelash** [comment]
Maybe we should also consider providing a non-JSON C API. A JSON API (or a binary that talks JSON RPC) could then still be built on top of the non-JSON C or the Rust API.
Are there any other projects providing a library that exposes a JSON C API?

**Comment by ArazAbishov** [comment]
As far as I understood from the conversation on irc channel, plugins are still supposed to run in separate processes. Do you have any ideas how then plugins could be used on platforms which do not allow sub-processes like iOS?

In general, do you have any plans on doing this refactor in near future?

Thanks.

**Comment by cmyr** [comment]
@ArazAbishov this is tentatively planned for early 2018.

For iOS, plugins will have to be able to compile as a static lib, and be bundled with the app. We haven't discussed an exact API for this, but it should be either the same or very similar to the eventual xilib API.

**Comment by ArazAbishov** [comment]
@cmyr That's pretty sweet. I am currently working on a prototype of xi-android using sub-processes and was curious if I should invest more time into it or just wait for refactoring to be finished. I guess the second option is the way to go then.

**Comment by cmyr** [comment]
@ArazAbishov fwiw, this change won't require a lot of work from frontend developers, so don't hold off because you're worried about having to throw away a lot of work. We'll probably provide an executable that wraps the library and functions just like the existing `xi-core`; and if you want to run in-process it's mostly just replacing a subprocess with a thread. Commands will still use the existing json-rpc protocol.

**Comment by unmellow** [comment]
since people aren't going to be hand editing the JSON 
files would it make sense to use BSON instead 
I'm no professional but I'm curious if there's any benefit
to using BSON in place of JSON

**Comment by cmyr** [comment]
@unmellow to date, the RPC mechanism has not been a performance bottleneck for us, so we haven't spent too much time evaluating alternatives.

**Comment by unmellow** [comment]
so you chose JSON because it's the first thing you thought of 
and there's no problem in it worth fixing fair enough
(and sorry if I sound rude I really don't mean too)

**Comment by cmyr** [comment]
@unmellow I mean more or less, except for the "first thing you thought of". JSON has the advantage that relatively high performance encoding/decoding is available on almost every platform, which is important when you want people to be able to write frontends. 

In any case, here are some good issues for people to offer all of their opinions on transport protocols (these opinions will continue to be ignored for the time being):
https://github.com/xi-editor/xi-editor/issues/514
https://github.com/xi-editor/xi-editor/issues/131

**Comment by unmellow** [comment]
yeah i know that part sounded rude but i couldn't think of a better way to say it and thanks for the links

**Comment by Cogitri** [comment]
I guess this is succeeded by #1168 ?
